framework/validators/EmailValidator.php

```diff
@@ -176,4 +176,25 @@ public function getClientOptions($model, $attribute)
 
         return $options;
     }
+
+    /**
+     * @param string $value
+     * @return string|bool returns string if it is valid and/or can be converted, bool false if it can't be converted and/or is invalid
+     * @see https://github.com/yiisoft/yii2/issues/18585
+     */
+    private function idnToAsciiWithFallback($value)
+    {
+        $ascii = $this->idnToAscii($value);
+        if ($ascii === false) {
+            $parts = explode('@', $this->pattern, 2);
+            $pattern = $parts[0] . "$/";
+            $fullPatternParts = explode('@', $this->fullPattern);
+            $fullPattern = $fullPatternParts[0] . '@' . $fullPatternParts[1]  ."$/";
```

Comment:
Why exploding and then imploding?

```suggestion
            $fullPattern = $this->fullPattern ."$/";
```

Reply:
The `pattern` variable contains validation for the full email address [part before the @, the part after the @] when `allowName` is false. 

The `fullPattern` variable is the same, but used when `allowName` is true. 

For this method, we just want to test the first part (the part before the @), not the second part . Using the unaltered regex would cause it to fail.

We could set a new regex, but for functionality reasons, and to hopefully not have to change this check should the regex change, I decided to split the regex and discard the check for the part after the @.

The `fullPattern` has an extra `@` in the beginning:  `/^[^@]...@...$/` vs the `pattern` variable: `/^...@...$/`.

I am by no means a regex expert, but it validates all the test cases like this. When splitting the string by the @ in explode, I needed to re-include it.

The idea being this: The regex used to validate the address works - in some cases, when passed into the ASCII conversion it fails, but ASCII conversion failing does not mean it is an invalid email address, nor does it mean it is a valid email address. So if it fails ASCII conversion in this method (which is just for the first part of the address), we still need to validate it, which we can use part of the regex (but not all of it) that already works to do that.

We could split the regex, as I have done, or tack a known good email address on the back and use the full regex. Or maybe something I didn't see.

If there is another way, open to suggestions. I don't this suggestion would work, however.

Reply:
The other option would be to take all validation out of this method. If the ASCII conversion results in false, just set it at the original value as it is validated with regex later on.

If that is the case, keeping the ASCII conversion would be done to support backwards compatibility.

Reply:
Since `$this->fullPattern` is subject to change by the configuration the whole check can fail with an error. Isn't it better to provide a new pattern for this only?

Reply:
@bizley assumptions about the pattern value are alright.

Reply:
The regex pattern could change (I believe I noticed it hasn't changed in several years, but that doesn't preclude it from future change). I put this in place with that in mind, actually. The purpose of this was to not have to maintain two sets of variables (one set being just a subset of another), as this could capture (most of) the changes without having to then understand and/or change a second related variable.

If there are changes in either of these patterns, the same changes would then have to be subsequently reflected in additional private variables.

The pattern would have to fundamentally change the paradigm it is using (i.e. test a set values, require an `@` symbol test another set of values), , not just the content of its check (i.e. the regex for each specific set of values), to break this. At that point, a refactor of several methods, including this one, would be required.

In that respect, it isn't really future proofing yourself or providing any real extra protection, in my opinion (which is why I coded it this way).

If you prefer, I can set these are new private variables and just leave a note that they should be updated whenever the regex to validate the full address is used.

Reply:
It's not about the pattern age here, nothing stops user from setting `->fullPattern = 'lalala'`, since we have these patterns public. We could protect them but it's too late for this now. The solution is to add another pattern so you will not have to operate on the original one, you will only have to use it.

Reply:
Yeah, that's fine. I will make the changes and re-push the branch.